can/interfaces/ics_neovi/neovi_bus.py

```diff
@@ -118,18 +149,28 @@ def __init__(self, channel, can_filters=None, **kwargs):
         type_filter = kwargs.get('type_filter')
         serial = kwargs.get('serial')
         self.dev = self._find_device(type_filter, serial)
-        ics.open_device(self.dev)
 
-        if 'bitrate' in kwargs:
-            for channel in self.channels:
-                ics.set_bit_rate(self.dev, kwargs.get('bitrate'), channel)
+        with open_lock:
+            ics.open_device(self.dev)
 
-        fd = kwargs.get('fd', False)
-        if fd:
-            if 'data_bitrate' in kwargs:
+        try:
+            if "bitrate" in kwargs:
                 for channel in self.channels:
-                    ics.set_fd_bit_rate(
-                        self.dev, kwargs.get('data_bitrate'), channel)
+                    ics.set_bit_rate(self.dev, kwargs.get("bitrate"), channel)
+
+            if kwargs.get("fd", False):
+                if "data_bitrate" in kwargs:
+                    for channel in self.channels:
+                        ics.set_fd_bit_rate(
+                            self.dev, kwargs.get("data_bitrate"), channel
+                        )
+        except ics.RuntimeError as re:
+            logger.error(re)
+            err = ICSApiError(*ics.get_last_api_error(self.dev))
+            try:
+                self.shutdown()
```

Comment:
If there is no `except` clause then you can remove the `try` and `finally`.

Reply:
Without the try/finally the exception `err` from line 169 would be lost if there is a exception in `self.shutdown(). Here the interesting exception is the one from line 169.

Reply:
If an exception happens in `bus.shutdown()` it is raised immediately since there is no `except` clause. Am i missing something?

Reply:
You can try this code bellow to see how this works:

```python
class ErrorOne(Exception):
    pass

class ErrorTwo(Exception):
    pass


def method_one():
    try:
        raise ErrorOne()
    finally:
        raise ErrorTwo()

def method_two():
    raise ErrorOne()

if __name__ == '__main__':
    try:
        method_one()
    except ErrorOne:
        print("Got Error One")
    except ErrorTwo:
        print("Got Error Two")

    try:
        method_two()
    except ErrorOne:
        print("Got Error One")
    except ErrorTwo:
        print("Got Error Two")
```

Reply:
Thank you for the example, i have never seen that before. I just found the [documentation](https://docs.python.org/3/tutorial/errors.html#defining-clean-up-actions) for it too.